File: bemkl_supervised_multilabel_classification_variational_test.m
function prediction = bemkl_supervised_multilabel_classification_variational_test(Km, state)
    N = size(Km, 2);
    P = size(Km, 3);
    L = length(state.be.mu) - P;

    prediction.G.mu = zeros(P, N, L);
    prediction.G.sigma = zeros(P, N, L);
    for o = 1:L
        for m = 1:P
            prediction.G.mu(m, :, o) = state.A.mu(:, o)' * Km(:, :, m);
            prediction.G.sigma(m, :, o) = state.parameters.sigma_g^2 + diag(Km(:, :, m)' * state.A.sigma(:, :, o) * Km(:, :, m));
        end
    end
    
    prediction.F.mu = zeros(L, N);
    prediction.F.sigma = zeros(L, N);
    for o = 1:L
        prediction.F.mu(o, :) = [ones(1, N); prediction.G.mu(:, :, o)]' * state.be.mu([o, L + 1:L + P]);
        prediction.F.sigma(o, :) = 1 + diag([ones(1, N); prediction.G.mu(:, :, o)]' * state.be.sigma([o, L + 1:L + P], [o, L + 1:L + P]) * [ones(1, N); prediction.G.mu(:, :, o)]);
    end    

    pos = 1 - normcdf((+state.parameters.margin - prediction.F.mu) ./ prediction.F.sigma);
    neg = normcdf((-state.parameters.margin - prediction.F.mu) ./ prediction.F.sigma);
    prediction.P = pos ./ (pos + neg);
end

File: bemkl_supervised_multilabel_classification_variational_test.R
# Mehmet Gonen (mehmet.gonen@gmail.com)

bemkl_supervised_multilabel_classification_variational_test <- function(Km, state) {
  N <- dim(Km)[2]
  P <- dim(Km)[3]
  L <- length(state$be$mu) - P

  G <- list(mu = array(0, c(P, N, L)), sigma = array(0, c(P, N, L)))
  for (o in 1:L) {
    for (m in 1:P) {
      G$mu[m,,o] <- crossprod(state$A$mu[,o], Km[,,m])
      G$sigma[m,,o] <- state$parameters$sigma_g^2 + diag(crossprod(Km[,,m], state$A$sigma[,,o]) %*% Km[,,m])
    }
  }
  
  F <- list(mu = matrix(0, L, N), sigma = matrix(0, L, N))
  for (o in 1:L) {
    F$mu[o,] <- crossprod(state$be$mu[c(o, (L + 1):(L + P))], rbind(matrix(1, 1, N), G$mu[,,o]))
    F$sigma[o,] <- 1 + diag(crossprod(rbind(matrix(1, 1, N), G$mu[,,o]), state$be$sigma[c(o, (L + 1):(L + P)), c(o, (L + 1):(L + P))]) %*% rbind(matrix(1, 1, N), G$mu[,,o]))
  }

  pos <- 1 - pnorm((+state$parameters$margin - F$mu) / F$sigma)
  neg <- pnorm((-state$parameters$margin - F$mu) / F$sigma)
  P <- pos / (pos + neg)

  prediction <- list(G = G, F = F, P = P)
}